### PR TITLE
Build Perl/Python bindings for subversion

### PR DESCRIPTION
**Pull request by KineticTheory**
This PR activate subversion's make targets `swig-py` and `swig-pl` to provide python and Perl bindings.
    
To build and install the Perl and python bindings for subversion, I needed to register new dependencies on swig, python and perl. The subversion build system requires that Makefile targets `swig-py`, `install-swig-py`, `swig-pl` and `install-swig-pl` to be called. Reference: [INSTALL](http://svn.apache.org/repos/asf/subversion/trunk/INSTALL)
    
For the systems I was working on, I also needed to register/use spack provided autoconf, automake, libtool, expat and pkg-config as build-only dependencies and I needed to call `./autogen.sh` from the source tree before running `configure`.

I'm no expert in building subversion, so I'm asking for help with testing and reviewing this PR.  For example, the way the python bindings work looks a bit like a python extension in spack.  Maybe my implementation should be modified to use that infrastructure. Should these bindings be presented as variants instead of defaults?

### CODE REVIEW

**Comment by KineticTheory** [comment]
I thought the `url = ` line was exempt from flake8 line-length requirements. I'm not sure how to fix this style issue.

**Comment by tgamblin** [comment]
@KineticTheory: style checker had a bug :)

**Comment by tgamblin** [comment]
I think it would make sense to make the various dependencies here optional, and to make this so that it `extends` `python` instead of just `depends_on`.  The latter will work assuming `subversion` installs its `python` stuff in `lib/pythonX.Y/site-packages`... does it?

**Comment by tgamblin** [comment]
@KineticTheory: FYI other people needed the flake8 fix, so I cherry-picked it onto `develop`, so you can just remove that commit from this branch and rebase on develop if this causes conflicts.

File: lib/spack/spack/cmd/flake8.py [hunk 1]
```diff
@@ -76,18 +76,17 @@
...
-    with open(source) as infile:
-        parent = os.path.dirname(dest)
-        mkdirp(parent)
 
-        with open(dest, 'w') as outfile:
+        with open(source) as infile:
+            parent = os.path.dirname(dest)
+            mkdirp(parent)
+
+            with open(dest, 'w') as outfile:
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/flake8.py]: Looks like there are still some changes to this file that need to be removed.

File: var/spack/repos/builtin/packages/ruby/package.py [hunk 2]
```diff
@@ -41,6 +41,7 @@ class Ruby(Package):
     depends_on('openssl', when='+openssl')
     variant('readline', default=False, description="Enable Readline support")
     depends_on('readline', when='+readline')
+    depends_on('binutils', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ruby/package.py]: In the docs we recommend against adding binutils as a dependency:
http://spack.readthedocs.io/en/latest/getting_started.html#binutils

**Comment by adamjstewart** [comment]
Why is the ruby support commented out? Is it not working?

File: var/spack/repos/builtin/packages/subversion/package.py [hunk 3]
```diff
@@ -33,51 +33,73 @@ class Subversion(Package):
...
+    # depends_on('ruby', when='+ruby')
 
     # Installation has race cases.
     parallel = False
 
+    # helper function to locate python's site-packages directory.
+    def python_sp_dir(self):
+        python_version = 'python' + self.spec['python'].version.up_to(2)
+        return join_path(spec['python'].prefix.lib, python_version,
+                         'site-packages')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/subversion/package.py]: If we need something like this, maybe it's better to put the code in the method `Python.setup_dependent_package`

Comment (tgamblin) [inline var/spack/repos/builtin/packages/subversion/package.py]: This should already be set up by the Python package.  You can get at it in dependencies of Python by referring to `self.site_packages_dir`.

Comment (alalazo) [inline var/spack/repos/builtin/packages/subversion/package.py]: @tgamblin is right. There's no need at all to have this function. If you use : 
```python
site_packages_dir = self.site_packages_dir
```
it should work. 

Furthermore, I think you have a typo in the current implementation : `python_sp_dir` is a method and it seems to me you call it like  a property.

**Comment by adamjstewart** [comment]
I would just uncomment the Ruby stuff. If it breaks, then someone else can fix it. But it won't break unless someone explicitly asks for it. And I'm sure it works.

File: var/spack/repos/builtin/packages/subversion/package.py [hunk 4]
```diff
@@ -33,51 +33,73 @@ class Subversion(Package):
     version('1.8.13',    '8065b3698d799507fb72dd7926ed32b6')
     version('1.9.3',     'a92bcfaec4e5038f82c74a7b5bbd2f46')
 
+    variant('python', default=False, description='Provide python bindings')
+    variant('perl', default=False, description='Provide Perl bindings')
+    # variant('ruby', default=False, description='Provide ruby bindings')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/subversion/package.py]: Good that these are False by default.  Language bindings typically bring in extra dependencies, and are not needed by the user who just wants to run `svn`.

File: var/spack/repos/builtin/packages/subversion/package.py [hunk 5]
```diff
@@ -33,51 +33,73 @@ class Subversion(Package):
...
+
     depends_on('apr')
     depends_on('apr-util')
     depends_on('zlib')
     depends_on('sqlite')
     depends_on('serf')
+    depends_on('expat')
+    depends_on('swig@:2.99')
+    depends_on('autoconf', type='build')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/subversion/package.py]: Why were these dependencies added unconditionally?  I presume `swig` is needed for the Python bindings.  These should have `when='+python'` or `when='+perl'` on them.  At the end of the day, this PR should add NO additional dependencies for the user who is not asking for Python or Perl bindings.

File: var/spack/repos/builtin/packages/subversion/package.py [hunk 6]
```diff
@@ -33,51 +33,73 @@ class Subversion(Package):
...
+    depends_on('automake', type='build')
+    depends_on('libtool', type='build')
+    depends_on('pkg-config', type='build')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/subversion/package.py]: Why are these needed?  Normally, tarball downloads come with the `configure` script already made, they don't require one to rerun `autogen.sh`.  What does this have to do with adding Perl + Python bindings?  If this is a "general improvement" of this package, it should go into a different PR, not piggy-backed onto P&P bindings.

File: var/spack/repos/builtin/packages/subversion/package.py [hunk 7]
```diff
@@ -33,51 +33,73 @@ class Subversion(Package):
...
 
-    # Optional: We need swig if we want the Perl, Python or Ruby
-    # bindings.
-    # depends_on('swig')
-    # depends_on('python')
-    # depends_on('perl')
-    # depends_on('ruby')
+    # Optional: We need swig if we want the Perl, Python or Ruby bindings.
+    extends('python', when='+python')
+    depends_on('perl@5.8:', when='+perl')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/subversion/package.py]: Please figure out whether the Python bindings work with Python2, Python3 or both; and add clauses accordingly.  If you need that, you should add a second line, eg: `depends_on('python@1.4:17.6')`

Also, does `when=` work on `extends('python')`?  It should, but I'm not sure if it does yet.

File: var/spack/repos/builtin/packages/subversion/package.py [hunk 8]
```diff
@@ -33,51 +33,73 @@ class Subversion(Package):
...
+        python_version = 'python' + self.spec['python'].version.up_to(2)
+        return join_path(spec['python'].prefix.lib, python_version,
+                         'site-packages')
+
     def install(self, spec, prefix):
 
+        # Pre-configure setup
+        sh = which('sh')
+        sh('./autogen.sh')
+
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/subversion/package.py]: There must be a better way than `which('sh')`.  Grep for "autogen" in other packages and copy what they do.
But I question the need to run autogen at all...

File: var/spack/repos/builtin/packages/subversion/package.py [hunk 9]
```diff
@@ -33,51 +33,73 @@ class Subversion(Package):
...
         options.append('--with-apr=%s' % spec['apr'].prefix)
         options.append('--with-apr-util=%s' % spec['apr-util'].prefix)
         options.append('--with-zlib=%s' % spec['zlib'].prefix)
         options.append('--with-sqlite=%s' % spec['sqlite'].prefix)
         options.append('--with-serf=%s' % spec['serf'].prefix)
-        # options.append('--with-swig=%s' % spec['swig'].prefix)
+        options.append('--with-expat=%s:%s:expat' %
+                       (spec['expat'].prefix.include,
+                        spec['expat'].prefix.lib))
+        options.append('--with-swig=%s' % spec['swig'].prefix)
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/subversion/package.py]: Is `expat` required when not building Python+Perl bindings?  If so, this should be inside an `if '+python' in spec` or similar block.

Same thing for Swig.. definitely should be in an if block.

**Comment by adamjstewart** [comment]
Ping @KineticTheory

**Comment by KineticTheory** [comment]
@adamjstewart I'm not going to have time to work on this for a few more weeks.  Should I close the PR?

**Comment by adamjstewart** [comment]
Nah, leave it open. I'll mark it up-for-grabs and someone can pick it up if they want to.

**Comment by tgamblin** [comment]
superseded by #3644
